hibernate-core/src/main/java/org/hibernate/engine/internal/Cascade.java

```diff
@@ -89,9 +94,11 @@ public static <T> void cascade(
 			final PersistenceContext persistenceContext = eventSource.getPersistenceContextInternal();
 			final EntityEntry entry = persistenceContext.getEntry( parent );
 			if ( entry != null
-					&& entry.getLoadedState() == null
+					&& ( ( entry.getLoadedState() == null
 					&& entry.getStatus() == Status.MANAGED
-					&& persister.getBytecodeEnhancementMetadata().isEnhancedForLazyLoading() ) {
+					&& persister.getBytecodeEnhancementMetadata().isEnhancedForLazyLoading() )
+					// No need to run check-on-flush for removed entities
+					|| action == CHECK_ON_FLUSH && entry.getStatus() == Status.DELETED ) ) {
```

Comment:
ahah, good catch, but .... is this really specific to `CHECK_ON_FLUSH`?

why do we *ever* need to process cascades for `DELETED` objects at flush time? Orphan delete maybe but....

what happens if you change it to:

    || entry.getStatus() == Status.DELETED
    
how many things break?

Reply:
The `entry` variable here is misleading because it's actually an instance of `EntityKey`. That's why I need the change inside the `cascade` method, where we already retrieved the entity entry instead.

Reply:
OK then, so something like:

```java
		for ( Map.Entry<Object,EntityEntry> me : persistenceContext.reentrantSafeEntityEntries() ) {
			final EntityEntry entry = me.getValue();
			if ( flushable(entry) ) {
				Cascade.cascade( CascadingActions.CHECK_ON_FLUSH, CascadePoint.BEFORE_FLUSH, session, entry.getPersister(), me.getKey(), null );
			}
		}
```

Reply:
Applied your suggestion, thanks.